Add non-ASCII character check for username suggestions

CodeWithEmad opened the pull request:
Added validation to prevent generating username suggestions for names containing non-ASCII characters. This fixes an issue where the system would generate usernames containing non-ASCII characters that would later be rejected by username validation.

Changes include:
- Added ASCII character validation in `generate_username_suggestions`
- Updated function's type hints and docstring
- Added comprehensive test cases covering ASCII, non-ASCII (including various scripts like Persian, Japanese, Greek, and Cyrillic), and edge cases

## Testing
- Added new test class `TestUsernameGeneration` with test cases for:
  - Special character removal
  - ASCII/non-ASCII handling
  - Uniqueness of suggestions
  - Maximum length compliance
  - Various edge cases

Close #35984

CodeWithEmad committed:
fix: Enhance username suggestion generation with ASCII validation and detailed docstring

- Updated `generate_username_suggestions` function to include validation for non-ASCII characters.
- Improved function documentation to clarify arguments, return types, and username generation logic.
- Added type hints for better code clarity and maintainability.

CodeWithEmad committed:
fix: Add username generation tests and utility functions

- Introduced `remove_special_characters_from_name` and `generate_username_suggestions` functions to enhance username handling.
- Added comprehensive test cases for username generation, including ASCII validation and uniqueness checks.
- Implemented tests for special character removal and suggestion generation based on various input scenarios, including edge cases.

CodeWithEmad committed:
chore: fixed typo

openedx-webhooks commented:
Thanks for the pull request, @CodeWithEmad!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

CodeWithEmad committed:
chore: Clean up workflow YAML files

- Removed unnecessary blank lines in the add-remove-label-on-comment.yml and check_python_dependencies.yml files to improve readability and maintainability.

CodeWithEmad commented:
Sorry to ping you here @kdmccormick but I noticed that it takes about 20-25 minutes to run all the lint/test jobs and lining is a big chunk of it (about 10-12 minutes!). Pylint is so slow and as the project grows, it takes more time just to check the linting and code style! Have we ever considered using [ruff](https://astral.sh/ruff)?

kdmccormick commented:
https://github.com/openedx/edx-platform/issues/32174

mphilbrick211 commented:
Hi @CodeWithEmad! Just checking to see if this is still in progress?

CodeWithEmad commented:
Hey @mphilbrick211 
I believe @kdmccormick reviewed the task, but the task  is stuck in the decision-making level where what exactly we should do for usernames with non-ascii characters. I mentioned the issue in the slack channels before to ask others opinions, but no luck there either.

kdmccormick commented on openedx/core/djangoapps/user_authn/utils.py:
```diff
@@ -72,12 +72,35 @@ def is_registration_api_v1(request):
+            Must contain only ASCII characters.
+
+    Returns:
+        list[str] | list: A list of up to 3 available username suggestions,
```
An empty list is a valid `list[str]` (because it vacuously it satisfies the predicate "every item in the list is a string"), so there is no need to specify `| list`. In fact, it is preferable to omit `| list`, because `list[str] | list` would allow for lists of integers, booleans, or any other type.

```suggestion
def generate_username_suggestions(name: str) -> list[str]:
    """
    Generate 3 available username suggestions based on the provided name.

    Args:
        name (str): The full name to generate username suggestions from.
            Must contain only ASCII characters.

    Returns:
        list[str]: A list of up to 3 available username suggestions,
```

kdmccormick commented on openedx/core/djangoapps/user_authn/tests/test_utils.py:
```diff
@@ -68,3 +72,73 @@ def test_safe_redirect_oauth2(self, client_redirect_uri, redirect_url, host, exp
+            assert remove_special_characters_from_name(input_name) == expected
+
+    @ddt.data(
+        # Test normal ASCII name
+        ('John Doe', True),  # Should return suggestions
+        ('Jane Smith', True),  # Should return suggestions
+        # Test non-ASCII names
+        ('José García', False),  # Contains non-ASCII characters
+        ('مریم میرزاخانی', False),  # Persian name
+        ('明美 田中', False),  # Japanese name
+        ('Σωκράτης', False),  # Greek name
+        ('Владимир', False),  # Cyrillic characters
+        # Test edge cases
+        ('A B', True),  # Minimal valid name
+        ('', True),  # Empty string
+        ('   ', True),  # Just spaces
```
🙌🏻

kdmccormick reviewed (approved):
Because this is a bugfix, I think it does not need product review. Please rebase and confirm that the feature works as expected on the PR sandbox (which should spin up within an hour). Please also change your commit message to `fix:`, since we are fixing an active bug rather than adding a new feature.

Thank you for finding and resolving this!

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36056-1021b6.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36056-1021b6.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9322294019-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9322294019-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9322294019-tutor-requirements.txt)

kdmccormick commented:
@CodeWithEmad ^ username/password for the sandbox is openedx/openedx

CodeWithEmad commented:
Thank you, dear @kdmccormick. 
I can confirm that the registration page doesn't suggest a username for non-ascii languages (Persian for the image)
![image](https://github.com/user-attachments/assets/8bcc990a-d656-4fe9-ab81-813ce12e0d7b)
and the original suggestion for ASCII names works as expected:
![image](https://github.com/user-attachments/assets/4e8e95b7-8f85-497c-83de-dfda7159ec0d)

kdmccormick reviewed (approved):
Excellent! I'll merge tomorrow.

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36056-1021b6.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36056-1021b6.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9324403898-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9324403898-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36056-1021b6-51564656-9324403898-tutor-requirements.txt)
